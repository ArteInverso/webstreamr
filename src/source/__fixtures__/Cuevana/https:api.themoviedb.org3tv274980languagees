{"adult":false,"backdrop_path":"/gSphqI0BhosjOz9DhWbqAEkowQh.jpg","created_by":[{"id":1879786,"credit_id":"68442fe17350568d43de674c","name":"David Zurdo","original_name":"David Zurdo","gender":0,"profile_path":null},{"id":236691,"credit_id":"68442fe92c399022f01ea04b","name":"Luis Marías","original_name":"Luis Marías","gender":2,"profile_path":null}],"episode_run_time":[],"first_air_date":"2025-06-13","genres":[{"id":18,"name":"Drama"},{"id":10759,"name":"Action & Adventure"}],"homepage":"","id":274980,"in_production":true,"languages":["es"],"last_air_date":"2025-06-13","last_episode_to_air":{"id":5695934,"name":"De piedra","overview":"El seguimiento a los terroristas falla: ahora tienen los explosivos y su plan sigue en marcha. Pero también sigue en marcha la investigación de Léon y Mario, que logran dar con el piso franco de los terroristas… aunque ellos ya no están allí. El atentado es inminente. ¿Podrán evitarlo?","vote_average":0.0,"vote_count":0,"air_date":"2025-06-13","episode_number":5,"episode_type":"finale","production_code":"","runtime":50,"season_number":1,"show_id":274980,"still_path":"/3ApMchNGajHWPPXLegjbjYF8FVE.jpg"},"name":"La frontera","next_episode_to_air":null,"networks":[{"id":1024,"logo_path":"/w7HfLNm9CWwRmAMU58udl2L7We7.png","name":"Prime Video","origin_country":""}],"number_of_episodes":6,"number_of_seasons":1,"origin_country":["ES"],"original_language":"es","original_name":"La frontera","overview":"Año 1987. Un grupo disidente de la banda terrorista ETA planea atentar en París. Solo tres personas, tan incompatibles como complementarias, podrán evitarlo.","popularity":4.4829,"poster_path":"/6SKzMXhp5q4yzk3f02QuNTQfv3M.jpg","production_companies":[{"id":240053,"logo_path":"/u6wqVYJTKPXc4yswTsoSdQLt2Oc.png","name":"Par Producciones","origin_country":"ES"}],"production_countries":[{"iso_3166_1":"ES","name":"Spain"}],"seasons":[{"air_date":"2025-06-13","episode_count":6,"id":424914,"name":"Temporada 1","overview":"Año 1987. Francia empieza a colaborar con España en la lucha contra ETA. Un grupo disidente de la organización va a cometer un atentado en París. Las autoridades españolas han resuelto no intervenir, pero un joven capitán de la Guardia Civil decide actuar. Una historia de acción, fidelidades y traiciones, donde cada uno tendrá que demostrar en qué cree y cuáles son sus “fronteras”.","poster_path":"/AkrrnHWs12H1my5HlI9dEVJS9a9.jpg","season_number":1,"vote_average":0.0}],"spoken_languages":[{"english_name":"Spanish","iso_639_1":"es","name":"Español"}],"status":"Returning Series","tagline":"","type":"Scripted","vote_average":0.0,"vote_count":0}