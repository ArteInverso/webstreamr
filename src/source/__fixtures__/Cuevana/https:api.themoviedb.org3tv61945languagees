{"adult":false,"backdrop_path":"/uYgcyrinlFfXTFkflzkOjAmtlpP.jpg","created_by":[{"id":49091,"credit_id":"54c38e099251416e6000effe","name":"Uli Brée","original_name":"Uli Brée","gender":2,"profile_path":"/8qHRQed5idzAlfrjc9j4FyNVttq.jpg"}],"episode_run_time":[45],"first_air_date":"2015-01-12","genres":[{"id":35,"name":"Comedia"},{"id":18,"name":"Drama"}],"homepage":"https://www.daserste.de/unterhaltung/serie/vorstadtweiber/index.html","id":61945,"in_production":false,"languages":["de"],"last_air_date":"2022-03-14","last_episode_to_air":{"id":3596974,"name":"Episodio 11","overview":"","vote_average":0.0,"vote_count":0,"air_date":"2022-03-14","episode_number":11,"episode_type":"finale","production_code":"","runtime":45,"season_number":6,"show_id":61945,"still_path":"/zsumt4d4TfvFB4QuXu2rp7jsaGl.jpg"},"name":"Vorstadtweiber","next_episode_to_air":null,"networks":[{"id":189,"logo_path":"/k1poBlW9HGDnIyZU67BtH3BXQuz.png","name":"ORF 1","origin_country":"AT"},{"id":308,"logo_path":"/nGl2dDGonksWY4fTzPPdkK3oNyq.png","name":"Das Erste","origin_country":"DE"}],"number_of_episodes":61,"number_of_seasons":6,"origin_country":["AT"],"original_language":"de","original_name":"Vorstadtweiber","overview":"","popularity":17.8497,"poster_path":"/um3A1r8A5Cuco7ldJHKtMhU7zU9.jpg","production_companies":[{"id":2279,"logo_path":null,"name":"MR Filmproduktion","origin_country":"AT"},{"id":6856,"logo_path":null,"name":"MR FILM","origin_country":"AT"}],"production_countries":[{"iso_3166_1":"AT","name":"Austria"}],"seasons":[{"air_date":"2015-03-31","episode_count":4,"id":74332,"name":"Especiales","overview":"","poster_path":"/tFF69cFAsIvWMYFp9NbxvQJAORE.jpg","season_number":0,"vote_average":0.0},{"air_date":"2015-01-12","episode_count":10,"id":64717,"name":"Temporada 1","overview":"","poster_path":"/aswimgrUIaeHPH3B8lPH2YYUZxj.jpg","season_number":1,"vote_average":9.0},{"air_date":"2016-03-14","episode_count":10,"id":74313,"name":"Temporada 2","overview":"","poster_path":"/7fVAfhzCXitqVVvvfMDzoNZ9uY1.jpg","season_number":2,"vote_average":9.0},{"air_date":"2018-01-08","episode_count":10,"id":97704,"name":"Temporada 3","overview":"","poster_path":"/W6kdpyhpYK6bU77OD3k3YXPH4k.jpg","season_number":3,"vote_average":9.0},{"air_date":"2019-09-16","episode_count":10,"id":132439,"name":"Temporada 4","overview":"","poster_path":"/hhQeOEgvbzzOBCybHISHQAPAHk6.jpg","season_number":4,"vote_average":8.0},{"air_date":"2021-01-11","episode_count":10,"id":171866,"name":"Temporada 5","overview":"","poster_path":"/5n93sCEdcSMyXc3WEJjJvENdztY.jpg","season_number":5,"vote_average":0.0},{"air_date":"2022-01-10","episode_count":11,"id":236440,"name":"Temporada 6","overview":"","poster_path":"/9GomC3oI57Tf3Y70WRMrpCs1GPg.jpg","season_number":6,"vote_average":0.0}],"spoken_languages":[{"english_name":"German","iso_639_1":"de","name":"Deutsch"}],"status":"Ended","tagline":"","type":"Scripted","vote_average":8.1,"vote_count":13}