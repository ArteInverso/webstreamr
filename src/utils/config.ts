import { Config } from '../types';

export const getDefaultConfig = (): Config => {
  return { multi: 'on', en: 'on' };
};

export const showExternalUrls = (config: Config): boolean => 'includeExternalUrls' in config;

export const hasMultiEnabled = (config: Config): boolean => 'multi' in config;
