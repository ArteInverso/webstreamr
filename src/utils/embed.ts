import { unpack } from 'unpacker';
import { cachedFetchText } from './fetch';
import { scanFromResolution } from './resolution';

const LINK_REGEXPS = [
  /sources:\[{file:"(.*?)"/, // dropload, supervideo
];

interface ParsePackedEmbedResult {
  url: string;
  resolution: string | undefined;
  size: string | undefined;
}

const findResolution = (html: string): string | undefined => {
  const match = html.match(/(\d{3,}x\d{3,}),/);

  return match && match[1] ? scanFromResolution(match[1]) : undefined;
};

const findSize = (html: string): string | undefined => {
  const sizeMatch = html.match(/([\d.]+) ?([GM]B)/);

  return sizeMatch && sizeMatch[1] && sizeMatch[2] ? `${sizeMatch[1]} ${sizeMatch[2]}` : undefined;
};

export const parsePackedEmbed = async (url: string): Promise<ParsePackedEmbedResult> => {
  const html = await cachedFetchText(url);

  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found on ${url}`);
  }

  const unpacked = unpack(evalMatch[0]);

  for (const linkRegexp of LINK_REGEXPS) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return {
        url: 'https://' + linkMatch[1].replace(/^(https:)?\/\//, ''),
        resolution: findResolution(html),
        size: findSize(html),
      };
    }
  }

  throw new Error(`Could not find a stream link on ${url}`);
};
