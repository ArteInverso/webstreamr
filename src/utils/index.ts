export * from './env';
export * from './Fetcher';
export * from './StreamResolver';
export * from './embed';
export * from './imdb';
export * from './languageFromCountryCode';
export * from './manifest';
export * from './tmdb';
