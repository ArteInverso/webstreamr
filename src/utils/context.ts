import { Request } from 'express';
import { Context } from '../types';
import { getDefaultConfig } from './config';

export const contextFromRequest = (req: Request): Context => {
  return {
    hostUrl: new URL(`${req.protocol}://${req.host}`),
    id: req.headers['X-Request-ID'] as string,
    ...(req.ip && { ip: req.ip }),
    config: req.params['config'] ? JSON.parse(req.params['config']) : getDefaultConfig(),
  };
};
