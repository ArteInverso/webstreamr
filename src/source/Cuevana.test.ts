import { createTestContext } from '../test';
import { FetcherMock, TmdbId } from '../utils';
import { Cuevana } from './Cuevana';

const ctx = createTestContext({ es: 'on', mx: 'on' });

describe('Cuevana', () => {
  let handler: Cuevana;

  beforeEach(() => {
    handler = new Cuevana(new FetcherMock(`${__dirname}/__fixtures__/Cuevana`));
  });

  test('handles non-existent series gracefully', async () => {
    const streams = await handler.handle(ctx, 'series', new TmdbId(61945, 1, 1));
    expect(streams).toHaveLength(0);
  });

  test('handles non-existent episodes gracefully', async () => {
    const streams = await handler.handle(ctx, 'series', new TmdbId(1402, 1, 99));
    expect(streams).toHaveLength(0);
  });

  test('handle walking dead s1e1', async () => {
    const streams = await handler.handle(ctx, 'series', new TmdbId(1402, 1, 1));
    expect(streams).toMatchSnapshot();
  });

  test('handle la frontera s1e1', async () => {
    const streams = await handler.handle(ctx, 'series', new TmdbId(274980, 1, 1));
    expect(streams).toMatchSnapshot();
  });

  test('handle el camino', async () => {
    const streams = await handler.handle(ctx, 'movie', new TmdbId(559969, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });

  test('does not return mx content for es', async () => {
    const streams = await handler.handle(createTestContext({ es: 'on' }), 'movie', new TmdbId(559969, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });

  test('does not return es content for mx', async () => {
    const streams = await handler.handle(createTestContext({ mx: 'on' }), 'movie', new TmdbId(559969, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });
});
