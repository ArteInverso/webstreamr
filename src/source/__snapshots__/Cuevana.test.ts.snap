// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Cuevana handle el camino 1`] = `
[
  {
    "countryCode": "mx",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://streamwish.to/e/9icce2ca0yir",
  },
  {
    "countryCode": "mx",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://filemoon.sx/e/fgxxkidqq4jp",
  },
  {
    "countryCode": "mx",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://voe.sx/e/h87m8wyj5x8j",
  },
  {
    "countryCode": "mx",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://doodstream.com/e/ju53ufkqy9hc",
  },
  {
    "countryCode": "mx",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://streamtape.com/e/WgxgBAwOyGfbZxY",
  },
  {
    "countryCode": "mx",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://waaw.to/f/xtG7M9GL2MmV",
  },
  {
    "countryCode": "mx",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://filelions.to/v/tnehp8g7zght",
  },
  {
    "countryCode": "mx",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://plustream.com/embedb2/gHHRwqKdHgF5MYniuxGWnG8j_vIW77uC4a4vrO5iwLktMwS8ryr2i.bhH3SiGHw6PKNBoqxCHLDyc1Pw6EPhwg--",
  },
  {
    "countryCode": "es",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://streamwish.to/e/mtb09rms1njz",
  },
  {
    "countryCode": "es",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://filemoon.sx/e/r77fu9ez9nsv",
  },
  {
    "countryCode": "es",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://voe.sx/e/cepsnqy0qgyo",
  },
  {
    "countryCode": "es",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://doodstream.com/e/ny5qlz61x9xt",
  },
  {
    "countryCode": "es",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://waaw.to/f/hlHhgFkmWHom",
  },
  {
    "countryCode": "es",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://plustream.com/embedb2/DpAUEfJrAHkBAdb.6erU25OAypGXqMugDfQuWU2sTDBKHNQmIxzyq1b2Ma06kb7EgL6d4hdzw8quWDimAGnIvQ--",
  },
  {
    "countryCode": "es",
    "title": "El Camino: Una película de Breaking Bad",
    "url": "https://vidhidepro.com/v/lhaueh479t2v",
  },
]
`;

exports[`Cuevana handle la frontera s1e1 1`] = `
[
  {
    "countryCode": "es",
    "title": "La frontera 1x1",
    "url": "https://streamwish.to/e/tpzbrzz7p41k",
  },
  {
    "countryCode": "es",
    "title": "La frontera 1x1",
    "url": "https://filemoon.sx/e/nhbmj6o6zl0l",
  },
  {
    "countryCode": "es",
    "title": "La frontera 1x1",
    "url": "https://vidhidepro.com/v/x5w6b0wys1fs",
  },
  {
    "countryCode": "es",
    "title": "La frontera 1x1",
    "url": "https://voe.sx/e/nfal9zqu9yvy",
  },
  {
    "countryCode": "es",
    "title": "La frontera 1x1",
    "url": "https://doodstream.com/e/dyvg7ofnnqgi",
  },
  {
    "countryCode": "es",
    "title": "La frontera 1x1",
    "url": "https://waaw.to/f/goypdL0qGqks",
  },
]
`;

exports[`Cuevana handle walking dead s1e1 1`] = `
[
  {
    "countryCode": "mx",
    "title": "The Walking Dead 1x1",
    "url": "https://streamwish.to/e/gudk6hvc4qwl",
  },
  {
    "countryCode": "mx",
    "title": "The Walking Dead 1x1",
    "url": "https://filemoon.sx/e/kp39yor1z0js",
  },
  {
    "countryCode": "mx",
    "title": "The Walking Dead 1x1",
    "url": "https://voe.sx/e/4ahrkbnofnty",
  },
  {
    "countryCode": "mx",
    "title": "The Walking Dead 1x1",
    "url": "https://doodstream.com/e/wvk23p39ikyd",
  },
  {
    "countryCode": "mx",
    "title": "The Walking Dead 1x1",
    "url": "https://streamtape.com/e/qMjrKleBMpCLO0",
  },
  {
    "countryCode": "mx",
    "title": "The Walking Dead 1x1",
    "url": "https://filelions.to/v/rvyogjvsejlh",
  },
]
`;
